src/Common/CurrentMemoryTracker.cpp

```diff
@@ -68,18 +68,24 @@ AllocationTrace CurrentMemoryTracker::allocImpl(Int64 size, bool throw_if_memory
         }
         current_thread->untracked_memory_blocker_level = blocker_level;
 
-        Int64 will_be = current_thread->untracked_memory + size;
-        if (will_be > current_thread->untracked_memory_limit)
+        size_t previous_untracked_memory = current_thread->untracked_memory;
+        current_thread->untracked_memory += size;
+        if (current_thread->untracked_memory > current_thread->untracked_memory_limit)
         {
-            auto res = memory_tracker->allocImpl(will_be, throw_if_memory_exceeded);
-            current_thread->untracked_memory = 0;
-            return res;
+            try
+            {
+                Int64 current_untracked_memory = current_thread->untracked_memory;
+                current_thread->untracked_memory = 0;
+                auto res = memory_tracker->allocImpl(current_untracked_memory, throw_if_memory_exceeded);
+                return res;
+            }
+            catch (...)
+            {
+                current_thread->untracked_memory = previous_untracked_memory;
```

Comment:
Should it be `+=`, in case `allocImpl` adjusted `untracked_memory` (by doing another alloc/free recursively) before throwing?

Reply:
Hm, you are right assignment is wrong, it should be actually `-= size` I think. Fixed.